pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -521,15 +529,36 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
 //  fqdn/NameManager instance).
 //  - Write the response to the endpoint.
 func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
-	stat := ProxyRequestContext{}
-	stat.ProcessingTime.Start()
 	requestID := request.Id // keep the original request ID
 	qname := string(request.Question[0].Name)
 	protocol := w.LocalAddr().Network()
 	scopedLog := log.WithFields(logrus.Fields{
 		logfields.DNSName:      qname,
 		logfields.IPAddr:       w.RemoteAddr(),
-		logfields.DNSRequestID: request.Id})
+		logfields.DNSRequestID: requestID})
+
+	var stat ProxyRequestContext
+	if p.ConcurrencyLimit != nil {
+		// TODO: Consider plumbing the daemon context here.
+		ctx, cancel := context.WithTimeout(context.TODO(), 1*time.Second)
+		defer cancel()
+
+		stat.SemaphoreAcquireTime.Start()
+		if err := p.ConcurrencyLimit.Acquire(ctx, 1); err != nil && errors.Is(err, context.DeadlineExceeded) {
+			err = fmt.Errorf(
+				"failed to acquire DNS proxy semaphore, %d parallel requests already in-flight: %w",
+				option.Config.DNSProxyConcurrencyLimit,
+				err,
+			)
+			scopedLog.WithError(err).Error("Dropping DNS request due to too many DNS requests already in-flight")
+			p.sendRefused(scopedLog, w, request)
+			return
+		}
+		stat.SemaphoreAcquireTime.End(true)
```

Comment:
The error condition here looks like it needs a few more steps to match the other error return cases in this function: update `stat.Err`, `p.NotifyOnDNSMsg()`. Furthermore the `stat.SemaphoreAcquireTime` is not currently being stopped in the error condition.

Reply:
Fixed